### ISSUE TITLE
require() loads same module multiple times if case is different

### ISSUE THREAD

**Issue by jessefulton** [nodejs/node-v0.x-archive#6829]
tl;dr - even though modules are frozen, a warning message could help debug some frustrating debug issues related to `require`'s lack of case sensitivity

At the risk of beating a dead horse... #5702, #6000, #2621

I just ended up debugging an issue for 2 hours (on OSX), turns out `require`'s lack of case sensitivity was the problem. Luckily I follow a naming convention and I spotted the capitalized filename in my own stack trace. Other than that, there was no clue as to what was going on. (FWIW, one file was `require`ing another using TitleCase instead of lowercase - this created two copies of the same module, which contained shared config settings.)

`require` should know on certain OS's that "./module.js" and "./Module.js" actually point to the same file and not create a separate instantiation of the module being required. IMHO, it should be case-sensitive so code is transportable. But at the very least, it could error/warn, or load in files in accordance with the OS, but it should store cached copies/references to the data loaded in based upon the _actual_ file name found.

Kind of a contrived example, but IMHO the following should either throw an error or work as expected, right now it simply breaks:

``` javascript
//config.js
module.exports.set = function(key, val) {/* */}
module.exports.get = function(key) {/* */}
```

``` javascript
//file1.js
var cfg = require("./config");
cfg.set("db", process.env.DB_CONNECTION_STRING);
```

``` javascript
//file 2.js
var cfg = require("./Config"); // <-- note capital 'C', this creates a NEW object, not the shared one as expected
var app = createApp();
app.connect(cfg.get("db")); // null
```

Since the second `cfg` was required using different case, a second object is returned and the "db" property hasn't been set and the app won't connect to the database.

**Comment by bdemirkir** [comment]
Main issue in here is the strange behaviour of fs.realpath. In many platforms realpath function actually gets the filename with correct case in case insensitive systems. But node does not. I'll bring this to io.js too. Maybe they decide to fix this.

**Comment by jasnell** [comment]
@orangemocha ... any further thoughts on this one?

**Comment by orangemocha** [comment]
There is a PR that has been open for a while but never made it through. https://github.com/joyent/node/pull/6774

I will revisit that PR, and build the case for it again. I will probably bring the PR over to io.js. If it gets approved there, we can decide if it makes sense to backport it.

**Comment by jasnell** [comment]
Ok, will mark this as defer-to-converge for now then.

**Comment by FransGH** [comment]
PR [#8145](https://github.com/nodejs/node-v0.x-archive/pull/8145) addresses this issue.
